crates/core/src/client/inner/event_loop.rs

```diff
@@ -0,0 +1,256 @@
+use std::sync::Arc;
+
+use futures::{channel::oneshot, pin_mut, select, FutureExt};
+use log::error;
+use phoenix_channels_client::{CallError, Event, EventPayload, Payload, PhoenixEvent, JSON};
+use tokio::sync::mpsc;
+
+use super::LiveViewClientState;
+use crate::{dom::ffi, error::LiveSocketError};
+
+pub struct LiveViewClientChannel {
+    /// Allows sending events back to the main event loop
+    message_sender: mpsc::UnboundedSender<ClientMessage>,
+}
+
+/// RPC to the main background event loop
+pub enum ClientMessage {
+    /// Send a message and wait for the response,
+    /// If it is an event it will be processed by the loop
+    Call {
+        response_tx: oneshot::Sender<Result<Payload, CallError>>,
+        event: Event,
+        payload: Payload,
+    },
+    /// Send a message and don't wait for a response
+    Cast { event: Event, payload: Payload },
+    /// Replace the current channel/
+    UpdateChannel,
+}
+
+pub(crate) struct EventLoop {
+    msg_tx: mpsc::UnboundedSender<ClientMessage>,
+    main_background_task: tokio::task::JoinHandle<()>,
+}
+
+impl EventLoop {
+    pub fn new(client_state: Arc<LiveViewClientState>) -> Self {
```

Comment:
I am not happy with the complexity of this function, it does quite a bit and has quite a few conditionals, it works and is pretty straightforward but frankly asks the reader to think too much about the application lifecycle. I think we should clean this up either now or in a future iteration.